### PR TITLE
[Perf] Optimize startup to not hit XmlSerializer path via LazyDeseria…

### PR DESCRIPTION
**Pull request by Therzok**
…lizer

### CODE REVIEW

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core.Assemblies/SystemAssemblyService.cs [hunk 1]
```diff
@@ -421,7 +421,7 @@ public TargetFrameworkMoniker GetTargetFrameworkForAssembly (TargetRuntime tr, s
 		void SaveUserAssemblyContext ()
 		{
 			List<string> list = new List<string> (userAssemblyContext.Directories);
-			PropertyService.Set ("MonoDevelop.Core.Assemblies.UserAssemblyContext", list);
```
Comment (alanmcgovern) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Core.Assemblies/SystemAssemblyService.cs]: Would it make sense to build this behaviour into the property service?

Comment (Therzok) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Core.Assemblies/SystemAssemblyService.cs]: So, I guess the check would be:
```
if (value is ICollection collection && collection.Count == 0)
    value = null;
```
?

Comment (Therzok) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Core.Assemblies/SystemAssemblyService.cs]: But yes, good idea.

Comment (alanmcgovern) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Core.Assemblies/SystemAssemblyService.cs]: yeah, something like that. If we do make this change we should also consider whether or not the deserializer should return an empty collection or a null value when the key is unset.

**Comment by KirillOsenkov** [comment]
I've been also thinking about optimizing the Loading side with this (in Properties.cs):

```
			public T Deserialize<T> ()
			{
				try {
					// optimize for a common case to avoid trying to load mscorlib.XmlSerializers.xml
					// for an empty List<string>
					if (typeof(T) == typeof(List<string>) &&
						xml != null &&
						xml.Contains(@"<ArrayOfString xmlns:xsi=""http://www.w3.org/2001/XMLSchema-instance"" xmlns:xsd=""http://www.w3.org/2001/XMLSchema"" />")) {
						return (T)(object)new List<string> ();
					}
```

It takes care of both cases where we try to read List<string>.

**Comment by alanmcgovern** [comment]
I don’t think we need to hardcode stuff like that into the loading side if we just fix the saving side. When those files are written to disk they just won’t have an empty array anymore, so the loader logic would become dead/unused code nearly instantly.

**Comment by Therzok** [comment]
@alanmcgovern @KirillOsenkov @mrward PTAL

**Comment by KirillOsenkov** [comment]
Thanks!

**Pull request merged**